### PR TITLE
[WIP] esp32/Makefile: Move to v4.1

### PR DESCRIPTION
**Pull request by osctobe**
Fixup build system for ESP-IDF v4.1. This only makes SDK parts build, there are API changes that need adjusting to in the code.

Issue #6412 .

### CODE REVIEW

File: ports/esp32/Makefile [hunk 1]
```diff
@@ -260,7 +265,7 @@ CFLAGS_COMMON = -Os -ffunction-sections -fdata-sections -fstrict-volatile-bitfie
 	-Wno-error=unused-variable -Wno-error=deprecated-declarations \
 	-DESP_PLATFORM
 
-CFLAGS_BASE = -std=gnu99 $(CFLAGS_COMMON) -DMBEDTLS_CONFIG_FILE='"mbedtls/esp_config.h"' -DHAVE_CONFIG_H
+CFLAGS_BASE = -std=gnu11 $(CFLAGS_COMMON) -DMBEDTLS_CONFIG_FILE='"mbedtls/esp_config.h"' -DHAVE_CONFIG_H
```
Comment (tve) [inline ports/esp32/Makefile:268]: Can you please explain the need for this change?

File: ports/esp32/Makefile [hunk 2]
```diff
@@ -398,17 +403,23 @@ $(SDKCONFIG_COMBINED): $(SDKCONFIG)
...
+	$(Q)$(CAT) $^ > $@
+
+$(ESPCOMP_KCONFIGS_PROJBUILD_SOURCE_FILE): $(ESPCOMP_KCONFIGS_PROJBUILD)
+	$(Q)$(CAT) $^ > $@
+
+$(SDKCONFIG_H): $(SDKCONFIG_COMBINED) $(ESPCOMP_KCONFIGS_SOURCE_FILE) $(ESPCOMP_KCONFIGS_PROJBUILD_SOURCE_FILE)
 	$(ECHO) "GEN $@"
 	$(Q)$(MKDIR) -p $(dir $@)
-	$(Q)$(PYTHON) $(ESPIDF)/tools/kconfig_new/confgen.py \
+	$(Q)$(PYTHON2) $(ESPIDF)/tools/kconfig_new/confgen.py \
```
Comment (tve) [inline ports/esp32/Makefile:415]: Why this change, seems something more related to your build env as opposed to esp-idf v4.0.1 vs. v4.1?

File: ports/esp32/Makefile [hunk 3]
```diff
@@ -562,7 +577,7 @@ ESPIDF_BT_NIMBLE_O = $(patsubst %.c,%.o,\
 endif
 
 $(BUILD)/$(ESPCOMP)/esp_eth/src/esp_eth_mac_dm9051.o: CFLAGS += -fno-strict-aliasing
-ESPIDF_ESP_ETH_O = $(patsubst %.c,%.o,$(wildcard $(ESPCOMP)/esp_eth/src/*.c))
+ESPIDF_ESP_ETH_O = $(patsubst %.c,%.o,$(filter-out %_openeth.c,$(wildcard $(ESPCOMP)/esp_eth/src/*.c)))
```
Comment (tve) [inline ports/esp32/Makefile:580]: Why does this have to be excluded?

File: ports/esp32/Makefile [hunk 4]
```diff
@@ -89,7 +91,7 @@ else ifeq ($(ESPIDF_CURHASH),$(ESPIDF_SUPHASH_V4))
 $(info Building with ESP IDF v4)
 
 PYPARSING_VERSION = $(shell python3 -c 'import pyparsing; print(pyparsing.__version__)')
-ifneq ($(PYPARSING_VERSION),2.3.1)
+ifeq ($(filter 2.3% 2.2% 2.1%,$(PYPARSING_VERSION)),)
```
Comment (tve) [inline ports/esp32/Makefile:94]: Is it really a good idea to allow all these older versions? Did you test against each one? Do we really want to test against each one?

**Review by tve — COMMENTED** [review]
Thanks, this is a great start! I did have a couple of questions, though...

**Comment by dpgeorge** [comment]
@osctobe thanks for the contribution here.  Are you able to keep working on this, or shall someone else pick it up?

File: ports/esp32/Makefile (hunk 1 above)
Comment (pumelo) [reply ports/esp32/Makefile:268]: I think this compile flag is only required for one file:

https://github.com/espressif/esp-idf/blob/94cb7e8b8f6da8abf2ca5c6117c81eaac92d90ee/components/driver/component.mk#L11

**Comment by pumelo** [comment]
This is more of a general question than related to the upgrade to v4.1:

Is there a specific reason that micropython does not build as a component to esp-idf?
If built as a component, this would lift the burden to maintain linker scripts and compile flags in micropython for esp-idf components.

See e.g. that the compile flags are changed for only one file https://github.com/espressif/esp-idf/blob/94cb7e8b8f6da8abf2ca5c6117c81eaac92d90ee/components/driver/component.mk#L11.

I know that there have been discussions about moving the esp32 port to cmake, but that's not what I'm talking about.

As long as esp-idf does support the "[legacy make](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-guides/build-system-legacy.html)" build system this should work quite well.

I have a working poc which builds micropython as a component to esp-idf with quite minimal change to the micropython code base. It uses the approch described here https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-guides/build-system-legacy.html#fully-overriding-the-component-makefile. I'll be happy to share this poc if interest is available.

**Comment by dpgeorge** [comment]
> Is there a specific reason that micropython does not build as a component to esp-idf?

Because it was done that way from the beginning, to have full control over the build, which is mainly to get the qstr generation done correctly.  It also makes it more like other ports.  But these are not strong reasons to do it that way.

> If built as a component, this would lift the burden to maintain linker scripts and compile flags in micropython for esp-idf components.

I agree.

> I know that there have been discussions about moving the esp32 port to cmake, but that's not what I'm talking about.
> 
> As long as esp-idf does support the "[legacy make](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-guides/build-system-legacy.html)" build system this should work quite well.

The legacy make support in the IDF may disappear in the future, so I'd say it's better to just go straight to cmake if anything.  See #6473 for a PR to do that.

**Comment by dpgeorge** [comment]
See #6906 .

**Comment by dpgeorge** [comment]
ESP IDF v4.1.1 is now supported, as of d191d88cabec9bf328f6f46b11ce6642af70cfc7

**Pull request closed without merging**